text/0000-pin.md

```diff
@@ -0,0 +1,408 @@
+trait Future {
+    type Item;
+    type Error;
+
+    fn poll(self: PinMut<Self>, cx: task::Context) -> Poll<Self::Item, Self::Error>;
+}
+```
+
+By default when implementing `Future` for a struct, this definition is equivalent to today's, which takes `&mut self`. But if you want to allow self-referencing in your future, you just opt out of `Unpin`, and `PinMut` takes care of the rest.
+
+This RFC also provides pinned analogies to `Box` and `&T` - `PinBox<T>` and `Pin<T>`. They work along the same lines as the `PinMut` type discussed here - if the type implements `Unpin`, they function the same as their unpinned varieties; if the type has opted out of `Unpin`, they guarantee that they type behind the reference will not be moved again.
+
+# Reference-level explanation
+
+## The `Unpin` auto trait
+
+This new auto trait is added to the `core::marker` and `std::marker` modules:
+
+```rust
+pub unsafe auto trait Unpin { }
+```
+
+A type that implements `Unpin` can be moved out of one of the pinned reference
+types discussed later in this RFC. Otherwise, they do not expose a safe API
+which allows you to move a value out of them. Because `Unpin` is an auto trait,
+most types in Rust implement `Unpin`. The types which don't are primarily
+self-referential types, like certain generators.
+
+This trait is a lang item, but only to generate negative impls for certain
+generators. Unlike previous `?Move` proposals, and unlike some traits like
+`Sized` and `Copy`, this trait does not impose any compiler-based semantics
+types that do or don't implement it. Instead, the semantics are entirely
+enforced through library APIs which use `Unpin` as a marker.
+
+## `Pin` and `PinMut`
+
+`Pin` and `PinMut` are two types added to `core::mem` and `std::mem`. They are
+analogous to `&T` and `&mut T` respectively.
+
+```rust
+#[fundamental]
+pub struct Pin<'a, T: ?Sized + 'a> {
+    data: &'a T,
+}
+
+#[fundamental]
+pub struct PinMut<'a, T: ?Sized + 'a> {
+    data: &'a mut T,
+}
+```
+
+### Safe APIs
+
+They both implement `Deref`, but `PinMut` only implements `DerefMut` if the
+type it references implements `Unpin`:
+
+```rust
+impl<'a, T: ?Sized> Deref for Pin<'a, T> { ... }
+
+impl<'a, T: ?Sized> Deref for PinMut<'a, T> { ... }
+
+impl<'a, T: Unpin + ?Sized> DerefMut for PinMut<'a, T> { ... }
+```
+
+They can only be safely constructed from references to types that implement
+`Unpin`:
+
+```rust
+impl<'a, T: Unpin + ?Sized> Pin<'a, T> {
+    pub fn new(reference: &'a T) -> Pin<'a, T> { ... }
+}
+
+impl<'a, T: Unpin + ?Sized> PinMut<'a, T> {
+    pub fn new(reference: &'a mut T) -> PinMut<'a, T> { ... }
+}
+```
+
+They also have a `borrow` function, which allows them to be transformed to pins
+with shorter lifetimes as may be necessary:
+
+```rust
+impl<'a, T: ?Sized> Pin<'a, T> {
+    pub fn borrow<'b>(this: &'b Pin<'a, T>) -> Pin<'b, T> { ... }
+}
+
+impl<'a, T: ?Sized> PinMut<'a, T> {
+    pub fn borrow<'b>(this: &'b PinMut<'a, T>) -> PinMut<'b, T> { ... }
```

Comment:
Should this really be a *shared* reference? Can't I use that to obtain two aliasing `PinMut`?

Reply:
it shoud not be a shared reference